### PR TITLE
Project: Push Notifications

### PR DESCRIPTION
**Pull request by simonrozsival**
## What's this?
This is the PN project PR: https://github.com/toggl/mobileapp/projects/60

## Why do we want this?
We want our apps to be synced with the server all the time.

## :squid: Permissions
🚫

### CODE REVIEW

File: Toggl.iOS/Services/PushNotificationsTokenServiceIos.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public sealed class PushNotificationsTokenServiceIos : IPushNotificationsTokenService
+    {
+        public PushNotificationsToken? Token => getToken();
+
+        public void InvalidateCurrentToken()
+        {
+            InstanceId.SharedInstance.DeleteId(error =>
+            {
+                Console.WriteLine("Error deleting instance");
```
Comment (heytherewill) [inline Toggl.iOS/Services/PushNotificationsTokenServiceIos.cs]: 👀

File: Toggl.iOS/Startup/AppDelegate.Notifications.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+using System;
```
Comment (heytherewill) [inline Toggl.iOS/Startup/AppDelegate.Notifications.cs]: Shouldn't this whole file be called `.PushNotifications` instead?

File: Toggl.iOS/Startup/AppDelegate.cs [hunk 3]
```diff
@@ -8,23 +8,31 @@
 using Toggl.Networking;
 using UIKit;
 using UserNotifications;
+using Firebase.CloudMessaging;
 
 namespace Toggl.iOS
 {
     [Register(nameof(AppDelegate))]
-    public sealed partial class AppDelegate : UIApplicationDelegate, IUNUserNotificationCenterDelegate
+    public sealed partial class AppDelegate : UIApplicationDelegate, IUNUserNotificationCenterDelegate, IMessagingDelegate
```
Comment (heytherewill) [inline Toggl.iOS/Startup/AppDelegate.cs]: Shouldn't the `IMessagingDelegate` be put in the file that implements it?

File: Toggl.Core/Interactors/PushNotifications/SubscribeToPushNotificationsInteractor.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                return Observable.Return(Unit.Default);
+
+            var previouslyRegisteredToken = keyValueStorage.GetString(PreviouslyRegisteredTokenKey);
+            var dateOfRegisteringTheToken = keyValueStorage.GetDateTimeOffset(DateOfRegisteringPreviousTokenKey);
+
+            if (!string.IsNullOrEmpty(previouslyRegisteredToken)
+                && previouslyRegisteredToken == token.ToString()
+                && dateOfRegisteringTheToken.HasValue
+                && !shouldReSubmitToken(dateOfRegisteringTheToken.Value))
+            {
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Core/Interactors/PushNotifications/SubscribeToPushNotificationsInteractor.cs]: I disagree. In the case of a condition spread over several lines, it's harder to spot where the condition ends and where the body of the if is. I'd like to keep it like this.

**Comment by simonrozsival** [comment]
There are now two (minor) conflicts between this branch and `develop`. They will be resolved with #5344

**Pull request merged**